week8.tex

```diff
@@ -48,4 +53,220 @@ \subsection{Statics}
+
+Finally, we also have two channels \kw{stdin} and \kw{stdout} which carry data of type \kw{string}.
+
+\subsubsection{Transitions}
+We present a subset of the state transitions:
+\todo{Needed to use \kw{s} in the step actions instead since $s$ shows up as $\int$ for some reason}
+\begin{mathpar}
+  \defrule[D-Input][dyn:input]
+  {\strut}
+  {\consEx*{s}{\cal I} \stepsTo<\sndAc*{\kw{stdin}}{\kw{s}}> {\cal I}}
+
+  \defrule[D-Output][dyn:output]
+  {\strut}
+  {{\cal O} \stepsTo<\rcvAc*{\kw{stdout}}{\kw{s}}> \consEx*{s}{\cal O}}
+
+  \defrule[D-Print][dyn:print]
+  {\strut}
+  {\kw{print}(s) \stepsTo<\sndAc*{\kw{stdout}}{\kw{s}}> {\retEx*{\unitEx*}}}
+
+  \defrule[D-Read][dyn:read]
+  {\strut}
+  {\kw{read} \stepsTo<\rcvAc*{\kw{stdin}}{\kw{s}}> {\retEx*{s}}}
+
+  \defrule[D-Single][dyn:single]
+  {S_1 \stepsTo<\alpha> S_1'}
+  {S_1 \parallel S_2 \stepsTo<\alpha> S_1' \parallel S_2}
+
+  \defrule[D-Par][dyn:par]
+  {S_1 \stepsTo<\alpha> S_1' \\
+    S_2 \stepsTo<\overline{\alpha}> S_2'}
+  {S_1 \parallel S_2 \stepsTo S_1' \parallel S_2'}
+
+\end{mathpar}
+Note that rule \ruleref{dyn:par} relies on the definition of $\overline{\alpha}$ representing the complementary action of $\alpha$. In this case, $\sndAc*{a}{s}$, the action to send $s$ over channel $a$, and $\rcvAc*{a}{s}$, the action to receive $s$ over channel $a$, are complementary to each other.
+
+The dynamics for I/O are relatively straightforward. If ${\cal I}$ is not empty and the program commands a \kw{read}, \ruleref{dyn:par} allows the overall configuration to take a step by pairing \ruleref{dyn:input} and \ruleref{dyn:read}. Similarly, if the program wishes to \kw{print}, \ruleref{dyn:output} and \ruleref{dyn:print} will apply concurrently.
+
+\ruleref{dyn:single} allows the program to take steps not pertaining to I/O according to the base CBPV dynamics. In this case $\alpha$ would be $\epsilon$, the silent action.
+
+\todo{Not sure how much of this to mention/is this extraneous detail? \ruleref{dyn:single} wasn't in the lecture, but I felt it was necessary to mention to show that the rest of the program can still run without needing to send messages}
+Note that this rule also allows for asynchronous sends since a process could perform an action $\sndAc*{a}{s}$ without a complementary receive. This is not an issue since we can view channels as a message queue, so any asynchronous send just adds its message to the queue. Reads must still be synchronous since there must be a value available on the queue to use in the rest of the program.
+
+\subsection{Equations}
+We present a few select equations:
+\begin{mathpar}
+
+  \defrule[E-Print-Fun][eq:print-fun]
+  {\Gamma \entails{\isOfTp{s}{\kw{string}}} \\
+    \Gamma, \isOfTp{x}{A} \entails{\isOfTp{C}{X}}}
```

Comment:
Are you sure using `\isOfTp x A` in the LHS of ⊢ is a good idea? 🤔 I sort of think the RHS using `\isOfTp` being a judgment and the LHS being a binding and hence they're different structures... But you are right that they use the same syntax. @HarrisonGrodin Any input?

Reply:
Not 100% sure, but I think this is what Bob usually does. (I agree it's kind of a different structure, but since it's syntactically the same, you can probably just leave it as-is.)